### ISSUE TITLE
Errors caused by generator or list comprehension in contract

### ISSUE THREAD

**Issue by treiher** [Parquery/icontract#127]
Using a generator or list comprehension in a contract seems not to be supported.

### Generator Comprehension

```Python
from icontract import invariant


@invariant(lambda self: all(" " not in n for n in self.parts))
class ID:
    def __init__(self, identifier: str) -> None:
        self.parts = identifier.split(".")


ID(" ")
```

```Console
Traceback (most recent call last):
  File "test.py", line 10, in <module>
    ID(" ")
  File "/home/tr/.local/lib/python3.8/site-packages/icontract/_checkers.py", line 435, in wrapper
    _assert_invariant(contract=contract, instance=instance)
  File "/home/tr/.local/lib/python3.8/site-packages/icontract/_checkers.py", line 175, in _assert_invariant
    msg = icontract._represent.generate_message(contract=contract, condition_kwargs={"self": instance})
  File "/home/tr/.local/lib/python3.8/site-packages/icontract/_represent.py", line 387, in generate_message
    repr_vals = repr_values(
  File "/home/tr/.local/lib/python3.8/site-packages/icontract/_represent.py", line 334, in repr_values
    recompute_visitor.visit(node=lambda_inspection.node.body)
  File "/usr/lib/python3.8/ast.py", line 360, in visit
    return visitor(node)
  File "/home/tr/.local/lib/python3.8/site-packages/icontract/_recompute.py", line 263, in visit_Call
    args.append(self.visit(node=arg_node))
  File "/usr/lib/python3.8/ast.py", line 360, in visit
    return visitor(node)
  File "/home/tr/.local/lib/python3.8/site-packages/icontract/_recompute.py", line 373, in visit_GeneratorExp
    result = self._execute_comprehension(node=node)
  File "/home/tr/.local/lib/python3.8/site-packages/icontract/_recompute.py", line 361, in _execute_comprehension
    code = compile(source=module_node, filename='<ast>', mode='exec')
TypeError: required field "posonlyargs" missing from arguments
```

### List Comprehension

```Python
from icontract import invariant


@invariant(lambda self: all([" " not in n for n in self.parts]))
class ID:
    def __init__(self, identifier: str) -> None:
        self.parts = identifier.split(".")


ID(" ")
```

```Console
Traceback (most recent call last):
  File "test.py", line 10, in <module>
    ID(" ")
  File "/home/tr/.local/lib/python3.8/site-packages/icontract/_checkers.py", line 435, in wrapper
    _assert_invariant(contract=contract, instance=instance)
  File "/home/tr/.local/lib/python3.8/site-packages/icontract/_checkers.py", line 175, in _assert_invariant
    msg = icontract._represent.generate_message(contract=contract, condition_kwargs={"self": instance})
  File "/home/tr/.local/lib/python3.8/site-packages/icontract/_represent.py", line 377, in generate_message
    decorator_inspection = inspect_decorator(lines=lines, lineno=condition_lineno, filename=filename)
  File "/home/tr/.local/lib/python3.8/site-packages/icontract/_represent.py", line 225, in inspect_decorator
    atok = asttokens.ASTTokens(decorator_text, parse=True)
  File "/home/tr/.local/lib/python3.8/site-packages/asttokens/asttokens.py", line 65, in __init__
    self.mark_tokens(self._tree)
  File "/home/tr/.local/lib/python3.8/site-packages/asttokens/asttokens.py", line 76, in mark_tokens
    MarkTokens(self).visit_tree(root_node)
  File "/home/tr/.local/lib/python3.8/site-packages/asttokens/mark_tokens.py", line 47, in visit_tree
    util.visit_tree(node, self._visit_before_children, self._visit_after_children)
  File "/home/tr/.local/lib/python3.8/site-packages/asttokens/util.py", line 192, in visit_tree
    ret = postvisit(current, par_value, value)
  File "/home/tr/.local/lib/python3.8/site-packages/asttokens/mark_tokens.py", line 90, in _visit_after_children
    nfirst, nlast = self._methods.get(self, node.__class__)(node, first, last)
  File "/home/tr/.local/lib/python3.8/site-packages/asttokens/mark_tokens.py", line 175, in visit_listcomp
    return self.handle_comp('[', node, first_token, last_token)
  File "/home/tr/.local/lib/python3.8/site-packages/asttokens/mark_tokens.py", line 171, in handle_comp
    util.expect_token(before, token.OP, open_brace)
  File "/home/tr/.local/lib/python3.8/site-packages/asttokens/util.py", line 56, in expect_token
    raise ValueError("Expected token %s, got %s on line %s col %s" % (
ValueError: Expected token OP:'[', got OP:'(' on line 1 col 28
```

### Versions

icontract  2.1.0
asttokens  1.1.15

**Comment by mristin** [comment]
Hi @treiher 
Thanks for reporting the issue. I'll try to have a look at it tomorrow. It seems like an issue with asttokens at first sight. Generators used to work :)

**Comment by mristin** [comment]
Indeed, this small snippet:
```python
        if int(platform.python_version_tuple()[1]) <= 7:
            func_def_node = ast.FunctionDef(
                name="generator_expr",
                args=ast.arguments(args=args, posonlyargs=[], kwonlyargs=[], kw_defaults=[], defaults=[]),
                decorator_list=[],
                body=[ast.Return(node)])

            module_node = ast.Module(body=[func_def_node], type_ignores=[])
        else:
            func_def_node = ast.FunctionDef(
                name="generator_expr",
                args=ast.arguments(args=args, posonlyargs=[], kwonlyargs=[], kw_defaults=[], defaults=[]),
                decorator_list=[],
                body=[ast.Return(node)])
```
fixed the first problem.

**Comment by mristin** [comment]
I am pretty sure that the second problem is directly related to asttokens. I created an issue at asttokens repository:
https://github.com/gristlabs/asttokens/issues/57

Using Python 3.6 would be a quick work-around till this is fixed if that is an option for you.

**Comment by treiher** [comment]
Thanks for the quick analysis! I can confirm that this issue is related to Python 3.8. I just tried using Python 3.7 and the problem did not occur.

**Comment by mristin** [comment]
Hi @treiher , I'd like to fix #126 before releasing the next patch version. That should happen soon (in the next couple of days).

**Comment by treiher** [comment]
That sounds good. Thanks!
